### PR TITLE
[v7r3] Supporting installing Python 3 clients with the SystemAdministrator

### PR DESCRIPTION
**Pull request by chrisburr**
Follow up to #5215. In principle I think this should be able to support switching between Python 2 and Python 3 installations on the same machine however changes will be needed to the local `bashrc` before installing a Python 3 version (will add to the wiki once the details are figured out).

I've tested this is working on the lbcertifdiracoauth machine, including for rolling back to the `old` version however there is one detail to still figure out. How will people specify the version to install?

The simplest option would be to allow people to put the DIRAC or VODIRAC version in the version field. The version of other DIRAC components (e.g. Vanilla DIRAC or WebAppDIRAC) cannot be constrained other than the package metadata and there is no ability to install unreleased code (e.g. `git+https//...`).

Alternatively the version field could be made a dictionary which can contain versions for multiple extensions, e.g. in LHCb on a machine which has `LHCbDIRAC`, `LHCbWebDIRAC`, and `VODIRAC` installed.

```python
version = {
    "LHCbDIRAC": "11.0.4",
    "WebAppDIRAC": "5.0.9",
    "VODIRAC": "git+https://github.com/DIRACGrid/VODIRAC.git@rel-5.0",
}
```

This would be converted into a pip install command that ensures all DIRAC packages that are currently installed are still installed, e.g.:

```bash
pip install \
    "LHCbDIRAC==11.0.4" \
    "DIRAC" \
    "git+https://github.com/DIRACGrid/VODIRAC.git@rel-5.0" \
    "LHCbWebDIRAC" \
    "WebAppDIRAC==5.0.9"
```

This is easy to add to the `SystemAdministratorHandler` however I'm not sure how best to expose it. I guess the update UI in the System Administrator page of the WebApp could be extended to show all currently installed extensions and offer a list of versions for each one (based on the PyPI API).

Any thoughts?

BEGINRELEASENOTES

*Framework
CHANGE: Extend Extension column in InstalledComponentsDB to be String(512)

*Python 3
NEW: The system administrator can now install Python 3 versions

ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py [hunk 1]
```diff
@@ -284,11 +291,101 @@ def export_addDefaultOptionsToCS(self, componentType, system, component, overwri
...
+        timeout=300,
+    )
+    if r.returncode != 0:
+      self.log.error(
+          "Installing DIRACOS2 failed with returncode",
+          "%s and stdout: %s" % (r.returncode, r.stderr)
+      )
+      return S_ERROR("Failed to install DIRACOS2 with message %s" % r.stderr)
+
+    # Update the pro link
```
Comment (fstagni) [inline src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py:375]: That's not a link anymore

File: src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py [hunk 2]
```diff
@@ -284,11 +291,101 @@ def export_addDefaultOptionsToCS(self, componentType, system, component, overwri
...
+    r = subprocess.run(
+        [
+            "%s/bin/pip" % installPrefix,
+            "install",
+            "--no-color",
+            # "--quiet",
+            # "DIRAC==%s" % version,
+            # "WebAppDIRAC,
+            "git+https://github.com/chrisburr/DIRAC.git@webapp-py3",
+            "git+https://github.com/chrisburr/WebAppDIRAC.git@python3-fixes-2",
```
Comment (fstagni) [inline src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py:361]: just reminder...

File: src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py [hunk 3]
```diff
@@ -13,9 +13,16 @@
...
+import tempfile
 
 import six
+from packaging.version import Version, InvalidVersion
+try:
+  import subprocess32 as subprocess
+except ImportError:
+  import subprocess
 
 # TODO: This should be modernised to use subprocess(32)
```
Comment (fstagni) [inline src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py:27]: remove this comment.

**Comment by fstagni** [comment]
> ```python
> version = {
>     "LHCbDIRAC": "11.0.4",
>     "WebAppDIRAC": "5.0.9",
>     "VODIRAC": "git+https://github.com/DIRACGrid/VODIRAC.git@rel-5.0",
> }
> ```

This looks a lot like releases.cfg

**Comment by chrisburr** [comment]
> This looks a lot like releases.cfg

It's different as this isn't a global hard coded file, the example I gave was of what you could hypothetically specify if you want to deploy something really specific.

Most of the time the call to update will have no manually specified arguments at all and it will just pick the latest patch release of everything.

File: src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py (hunk 3 above)
Comment (chrisburr) [reply src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py:27]: Why? This comment still applies as `commands` is used by `export_getHostInfo` and should be modernised at some point.

Comment (fstagni) [reply src/DIRAC/FrameworkSystem/Service/SystemAdministratorHandler.py:27]: a-ha, I missed that.

**Comment by chrisburr** [comment]
I think this is ready for review.

File: src/DIRAC/FrameworkSystem/DB/InstalledComponentsDB.py [hunk 4]
```diff
@@ -246,7 +246,7 @@ class HostLogging(componentsBase):
   hostName = Column('HostName', String(32), nullable=False, primary_key=True)
   # status
   DIRAC = Column('DIRACVersion', String(64))
-  Extension = Column('Extension', String(64))
+  Extension = Column('Extension', String(512))
```
Comment (fstagni) [inline src/DIRAC/FrameworkSystem/DB/InstalledComponentsDB.py:249]: This should be advertised in the wiki.

**Review by fstagni — DISMISSED** [review]

**Comment by atsareg** [comment]
There are conflicts with some earlier merged PRs. Please, rebase this one

**Pull request merged**